**Use Case**

A department has created a crazy new product which they believe will help their customers generate more revenue.
 
For this department revenue is linked to deposits that users make ie the higher the value deposited the better things are.

They are looking to determine if this crazy new product is doing what they hoped and if you can provide an estimate uptick or not of the value the product is generating.

In order for this entire project to be completely autonomous. We will use Python for the entirety of the Importing, Normalising & Wrangling of this Initial Excel Sheet, for when we get our next batch of data, the process will be repeated, with the same steps seamlessly and effortlessly. 

In [1]:
import pandas as pd
import numpy as np
import seaborn as sns
import plotly.express as px
%matplotlib inline
import matplotlib.pyplot as plt
from google.colab import files

In [3]:
uploaded = files.upload()

Saving New product Assessment_Dataset.xlsx to New product Assessment_Dataset.xlsx


We will save the unaltered dataframe as 'df1'

In [4]:
df1 = pd.read_excel('New product Assessment_Dataset.xlsx')
df1

UserAccountKey CustomerName   Website           County  \
0          86829017002    Customer1  website2  group countries   
1          90400831002    Customer1  website2  group countries   
2          94257600002    Customer1  website2  group countries   
3          94551460002    Customer1  website2  group countries   
4          96911609002    Customer1  website2  group countries   
...                ...          ...       ...              ...   
138529    685870858002    Customer1  website8        Country 1   
138530    685870859002    Customer1  website7        Country 1   
138531    685870860002    Customer1  website8        Country 1   
138532    685870862002    Customer1  website9        Country 1   
138533    685870868002    Customer1  website9        Country 1   

        iscrazynewproduct     Revenue  Amount Deposited  NumDeposits  
0                       0   -0.580000          0.000000            0  
1                       0  484.990000       1235.000000           25  
2                       0   -5.740000         50.000000            1  
3                       0  156.449864        170.820000           18  
4                       0  586.702507        842.581543           10  
...                   ...         ...               ...          ...  
138529                  0  -34.426220        172.131102            1  
138530                  0   34.412450          0.000000            0  
138531                  0   27.534091         27.540976            2  
138532                  0   34.426220         34.426220            1  
138533                  0   41.242612         41.311464            1  

[138534 rows x 8 columns]

In [5]:
print('Our initial dataset has a total of {} rows and a total of {} columns.'.format(len(df1.index),len(df1.columns)))

Our initial dataset has a total of 138534 rows and a total of 8 columns.


Lets check and remove any duplicate and non entries that might corrupt our analysis by using our 'UserAccountKey' as a key.We will also fix any spelling errors on column names. 

In [6]:
df1.rename(columns = {'County':'Country'}, inplace = True)
df1.drop_duplicates(subset=['UserAccountKey'])
df1.dropna(subset=['UserAccountKey'])

UserAccountKey CustomerName   Website          Country  \
0          86829017002    Customer1  website2  group countries   
1          90400831002    Customer1  website2  group countries   
2          94257600002    Customer1  website2  group countries   
3          94551460002    Customer1  website2  group countries   
4          96911609002    Customer1  website2  group countries   
...                ...          ...       ...              ...   
138529    685870858002    Customer1  website8        Country 1   
138530    685870859002    Customer1  website7        Country 1   
138531    685870860002    Customer1  website8        Country 1   
138532    685870862002    Customer1  website9        Country 1   
138533    685870868002    Customer1  website9        Country 1   

        iscrazynewproduct     Revenue  Amount Deposited  NumDeposits  
0                       0   -0.580000          0.000000            0  
1                       0  484.990000       1235.000000           25  
2                       0   -5.740000         50.000000            1  
3                       0  156.449864        170.820000           18  
4                       0  586.702507        842.581543           10  
...                   ...         ...               ...          ...  
138529                  0  -34.426220        172.131102            1  
138530                  0   34.412450          0.000000            0  
138531                  0   27.534091         27.540976            2  
138532                  0   34.426220         34.426220            1  
138533                  0   41.242612         41.311464            1  

[138534 rows x 8 columns]

In [7]:
print('After dropping duplicated entries our dataset has a total of {} rows and a total of {} columns.'.format(len(df1.index),len(df1.columns)))

After dropping duplicated entries our dataset has a total of 138534 rows and a total of 8 columns.


Our primary objective is to draw a comparison between customers that tried our crazy new product and those that did not try our new product. 
So we shall split the dataframe into those that tried the product and those that did not try the product into dataframes named: 'df2' and 'df3' respectively.

In [8]:
df2 = df1.loc[df1['iscrazynewproduct'] == 1]
df3 = df1.loc[df1['iscrazynewproduct'] == 0]

In [9]:
print('Out of a total pool of {} customers. It seems that only {} customers tried our new product and {} did not try our new product.'.format(len(df1.index),len(df2.index),len(df3.index)))

Out of a total pool of 138534 customers. It seems that only 995 customers tried our new product and 137539 did not try our new product.


There are also customers that tried the new product, but did not make any deposits. We would like to compare those numbers with customers who did not try the product but also did not make any deposits. splitting the dataframes again to those that tried and made deposits: 'df2_deposits' and those that tried but did not make deposits: 'df2_no_deposits'.

In [10]:
df2_deposits = df2.loc[df2['Amount Deposited'] != 0]
df2_no_deposits = df2.loc[df2['Amount Deposited'] == 0]

In [124]:
df2['NumDeposits'].value_counts()

1      53
3      51
4      44
5      42
2      40
       ..
83      1
71      1
96      1
129     1
89      1
Name: NumDeposits, Length: 108, dtype: int64

In [129]:
df3['NumDeposits'].value_counts()

0      57795
1      28888
2      12598
3       7500
4       4982
       ...  
149        1
190        1
217        1
153        1
233        1
Name: NumDeposits, Length: 197, dtype: int64

In [100]:
print('Out of a total pool of {} customers that tried the product, {} made deposits and {} made no deposits'.format(len(df2.index),len(df2_deposits),len(df2_no_deposits)))

Out of a total pool of 995 customers that tried the product, 960 made deposits and 35 made no deposits


We will do a similar split as the one before, but this time for customers that did not try the product. Using 'df3_deposit' and 'df3_no_deposit'.

In [13]:
df3_deposits = df3.loc[df3['Amount Deposited'] != 0]
df3_no_deposits = df3.loc[df3['Amount Deposited'] == 0]

In [14]:
print('Out of a total pool of {} customers that did not try the product, {} made deposits and {} made no deposits.'.format(len(df3),len(df3_deposits),len(df3_no_deposits)))

Out of a total pool of 137539 customers that did not try the product, 79744 made deposits and 57795 made no deposits.


In [15]:
tried_deposit = len(df2_deposits)
tried_no_deposit = len(df2_no_deposits)
no_try_deposit = len(df3_deposits)
no_try_no_deposit = len(df3_no_deposits)

In [16]:
print(tried_deposit)

960


In [17]:
names = ['Tried', 'Did Not Try']
data = [{'Deposit': tried_deposit, 'No Deposit': tried_no_deposit},
        {'Deposit': no_try_deposit, 'No Deposit': no_try_no_deposit}]
  
df = pd.DataFrame(data, index = names)
df

Deposit  No Deposit
Tried            960          35
Did Not Try    79744       57795

In [18]:
df = pd.DataFrame({
    "Tried": [tried_deposit, tried_no_deposit],
})

fig = px.pie(df, values="Tried", title="Deposit vs No Deposit on Customers that TRIED the crazy new product (Blue = Made Deposits)")
fig.show()


In [127]:
df = pd.DataFrame({
    "Did Not Try": [no_try_deposit, no_try_no_deposit],
})

fig = px.pie(df, values="Did Not Try", title="Deposit vs No Deposit on Customers that DID NOT TRY the crazy new product (Blue = Made Deposits)")
fig.show()

Our crazy new product seems to be a success, a whopping 96.5 % of customers that do try the product, make deposits! compared to only 58 % that did not try the product and make deposits. 

Who is most likely to try this crazy new product?
Assuming that the revenue is the indication on whether the company has made a profit on a customer or not. We will compare customers who've lost the company money (negative revenue) and customers who the company has made money on. To see if there are no differences between the two. 

In [25]:
df2_deposits['Revenue'].describe()

count      960.000000
mean       418.834083
std       1078.070921
min      -6954.744251
25%         54.670059
50%        214.316503
75%        588.922674
max      10045.580096
Name: Revenue, dtype: float64

In [36]:
df_no_outliers = df2_deposits

# Get the `Revenue` column
revenue = df_no_outliers['Revenue']

# Calculate the interquartile range (IQR)
iqr = np.quantile(revenue, 0.75) - np.quantile(revenue, 0.25)

# Find the lower and upper bounds
lower_bound = np.quantile(revenue, 0.25) - 1.5 * iqr
upper_bound = np.quantile(revenue, 0.75) + 1.5 * iqr

# Identify the outliers
outliers = revenue[(revenue < lower_bound) | (revenue > upper_bound)]

# Remove the outliers
df_no_outliers = df_no_outliers[df_no_outliers['Revenue'].isin(outliers) == False]

In [48]:
fig = px.box(df_no_outliers, x='Revenue')
df_no_outliers['Revenue'].describe()
fig.show()

In [47]:
df_no_outliers['Revenue'].describe()

count     843.000000
mean      278.890696
std       357.206083
min      -737.954280
25%        54.919756
50%       192.302417
75%       444.376215
max      1372.789684
Name: Revenue, dtype: float64

In [51]:
df2_no_outliers = df2_no_deposits

# Get the `Revenue` column
revenue = df2_no_outliers['Revenue']

# Calculate the interquartile range (IQR)
iqr = np.quantile(revenue, 0.75) - np.quantile(revenue, 0.25)

# Find the lower and upper bounds
lower_bound = np.quantile(revenue, 0.25) - 1.5 * iqr
upper_bound = np.quantile(revenue, 0.75) + 1.5 * iqr

# Identify the outliers
outliers = revenue[(revenue < lower_bound) | (revenue > upper_bound)]

# Remove the outliers
df2_no_outliers = df2_no_outliers[df2_no_outliers['Revenue'].isin(outliers) == False]

In [52]:
fig = px.box(df2_no_outliers, x='Revenue')

fig.show()

In [139]:
df2_no_outliers['Revenue'].describe()

count    32.000000
mean     16.145990
std      19.006046
min     -27.508906
25%       6.750678
50%      10.445595
75%      26.473021
max      60.367869
Name: Revenue, dtype: float64

In [110]:
import plotly.express as px

# Create the data
x_values = ['Tried & Made Deposit', 'Tried But Did Not Make A Deposit']
y_values = [round(df_no_outliers['Revenue'].mean()), round(df2_no_outliers['Revenue'].mean())]

# Create the bar chart
fig = px.bar(x=x_values, y=y_values,text=y_values)

# Show the chart
fig.show()

In [54]:
print('The company makes an average of {} USD on the customers that do try the product & make deposits, and an average of {} USD on customers that do not make deposits.'.format(round(df_no_outliers['Revenue'].mean()),round(df2_no_outliers['Revenue'].mean())))

The company makes an average of 279 USD on the customers that do make deposits, and an average of 16 USD on customers that do not make deposits.


The company makes more money on the customers that try the product and make deposits make deposits compared to the customers that try but do not make deposits.

We would also like to know which countries contribute the most customers that tried our product.

In [143]:
df2_deposits_no_outliers = df2_deposits

# Get the `Revenue` column
revenue = df2_deposits_no_outliers['Amount Deposited']

# Calculate the interquartile range (IQR)
iqr = np.quantile(revenue, 0.75) - np.quantile(revenue, 0.25)

# Find the lower and upper bounds
lower_bound = np.quantile(revenue, 0.25) - 1.5 * iqr
upper_bound = np.quantile(revenue, 0.75) + 1.5 * iqr

# Identify the outliers
outliers = revenue[(revenue < lower_bound) | (revenue > upper_bound)]

# Remove the outliers
df2_deposits_no_outliers = df2_deposits_no_outliers[df2_deposits_no_outliers['Amount Deposited'].isin(outliers) == False]

In [144]:
df2_deposits_no_outliers['Amount Deposited'].describe()

count     868.000000
mean      443.141048
std       427.012993
min         5.957079
25%       120.808775
50%       295.089627
75%       631.638922
max      1847.970889
Name: Amount Deposited, dtype: float64

In [145]:
df3_deposits_no_outliers = df3_deposits

# Get the `Revenue` column
revenue = df3_deposits_no_outliers['Amount Deposited']

# Calculate the interquartile range (IQR)
iqr = np.quantile(revenue, 0.75) - np.quantile(revenue, 0.25)

# Find the lower and upper bounds
lower_bound = np.quantile(revenue, 0.25) - 1.5 * iqr
upper_bound = np.quantile(revenue, 0.75) + 1.5 * iqr

# Identify the outliers
outliers = revenue[(revenue < lower_bound) | (revenue > upper_bound)]

# Remove the outliers
df3_deposits_no_outliers = df3_deposits_no_outliers[df3_deposits_no_outliers['Amount Deposited'].isin(outliers) == False]

In [146]:
df3_deposits_no_outliers['Amount Deposited'].describe()

count    69158.000000
mean        76.922805
std         94.470082
min          1.730284
25%         11.619432
50%         34.507804
75%        103.808954
max        431.906186
Name: Amount Deposited, dtype: float64

In [153]:
x_values = ['Tried & Average Amount Deposited(in USD)', 'Not Tried Average Amount Deposited(in USD)']
y_values = [round(df2_deposits_no_outliers['Amount Deposited'].mean()), round(df3_deposits_no_outliers['Amount Deposited'].mean())]

# Create the bar chart
fig = px.bar(x=x_values, y=y_values,text=y_values)

# Show the chart
fig.show()

In [156]:
print('The company makes an average of {} USD per amount deposited on the customers that do try the crazy new product, compared to {} USD per amount deposited on the users that did not try the product. '.format(round(df2_deposits_no_outliers['Amount Deposited'].mean()), round(df3_deposits_no_outliers['Amount Deposited'].mean())))

The company makes an average of 443 USD per amount deposited on the customers that do try the crazy new product, compared to 77 USD per amount deposited on the users that did not try the product. 


In [55]:
country_list = df1.Country.unique().tolist()
country_list.sort()

In [56]:
tried = []
for country in country_list:
  i = df1.loc[df1['Country'] == country]
  i = (i['iscrazynewproduct'] == 1).sum()
  tried.append(i)

In [57]:
not_tried = []
for country in country_list:
  j = df1.loc[df1['Country'] == country]
  j = (j['iscrazynewproduct'] == 0).sum()
  not_tried.append(j)

In [58]:
df_country = pd.DataFrame({
    'Country': country_list,
    'Tried Crazy New Product': tried,
    'Did Not Try Crazy New Product': not_tried
})

In [101]:
df_country

Country  Tried Crazy New Product  Did Not Try Crazy New Product
0        Country 1                      273                          12121
1        Country 2                        1                           3411
2  group countries                      721                         122007

In [60]:
df_country.iat[0,1]
df = pd.DataFrame({
    "Country 1": [df_country.iat[0,1], df_country.iat[0,2]],
})

fig = px.pie(df, values="Country 1", title="Country 1's Percentage of Customers Who Tried The Crazy New Product (Red = Tried)")
fig.show()

In [61]:
df_country.iat[0,1]
df = pd.DataFrame({
    "Country 2": [df_country.iat[1,1], df_country.iat[1,2]],
})

fig = px.pie(df, values="Country 2", title="Country 2's Percentage of Customers Who Tried The Crazy New Product (Red = Tried)")
fig.show()

In [86]:
df_country.iat[0,1]
df = pd.DataFrame({
    "Group Countries": [df_country.iat[2,1], df_country.iat[2,2]],
})

fig = px.pie(df, values="Group Countries", title="Group Countries' Percentage of Customers Who Tried The Crazy New Product (Red = Tried)")
fig.show()

Group 2 had only one customer that tried the crazy new product, while group countries has the largest number of customers that tried the crazy new product

Now finally, lets see which website contributes the most users that tried the crazy new product.

In [64]:
website_list = df1.Website.unique().tolist()
website_list.sort()

['website1', 'website2', 'website3', 'website4', 'website5', 'website6', 'website7', 'website8', 'website9']


In [66]:
tried_w = []
for website in website_list:
  i = df1.loc[df1['Website'] == website]
  i = (i['iscrazynewproduct'] == 1).sum()
  tried_w.append(i)

In [75]:
not_tried_w = []
for website in website_list:
  j = df1.loc[df1['Website'] == website]
  j = (j['iscrazynewproduct'] == 0).sum()
  not_tried_w.append(j)

In [76]:
df_website = pd.DataFrame({
    'Website': website_list,
    'Tried Crazy New Product': tried_w,
    'Did Not Try Crazy New Product': not_tried_w
})

In [89]:
df_website

Website  Tried Crazy New Product  Did Not Try Crazy New Product
0  website1                       14                           1605
1  website2                       49                          10791
2  website3                      306                          23396
3  website4                      352                          86215
4  website5                        1                           3411
5  website6                        6                            371
6  website7                      106                           3057
7  website8                       17                           1078
8  website9                      144                           7615

In [114]:
import plotly.express as px

fig = px.bar(df_website, x="Website", y="Tried Crazy New Product",text="Tried Crazy New Product", color="Website", title="Number Of Users That Tried The Crazy New Product Per Website")
fig.show()

In [116]:
fig = px.bar(df_website, x="Website", y="Did Not Try Crazy New Product",text="Did Not Try Crazy New Product", color="Website", title="Number Of Users Did Not Try The Crazy New Product Per Website")
fig.show()

From our final observation, it is quite clear that website4 has the largest contribution of users, both that tried the crazy new product and that did not try the crazy new product.

Website 5 has the least contribution of customers that tried and Website 6 has the least number of customers that did not try.

**Conclusion**

Considering how 96.5 % of users that tried the product make deposits compared to only 58 % of deposits for users that did not try the product. And how on average, the product generates 443 USD amount deposited compared to only 77 USD on the old product. I would consider the product to be a success. There are signs of early adoption and financial commitment from the users. Another data point that could help with analysis is the time each user spends on the product (retention) 